### PR TITLE
First Prototype of push-Pull-Compare method

### LINKED ISSUE #4: Test_Engine: Add method for push-pull adapter testing
Should add a method for basic level interop testing.

Idea of the method to have inputs:

- Adapter (Adapter to be tested)
- Type (Type of objects to be tested)
- Some form of config
- boolean to set it running

Method should

1. Grab all available data of the type from the TestSets in the DataSets
1. Filter out appropriate sets with the config (this can be added at a later stage)
1. Push all the objects, one set at the time, to the adapter
1. Pull back all the objects of the type from the adapter
1. Compare the two
1. Report back:
   - Did the push work (did the adapter think it managed to push the obejcts?)
   - Did the pull work (did you get any objects back)
   - Was the object equal (from comparison)
   - What was the difference before and after

The results reported back can then be stored on a server, and every time a PR is raced for an adapter, the CI system could run through this test for all applicable object types for the adapter and and report back if things are better/same/worse than before. As soon as the PR is merged this new difference can then be stored on the server as the new checking point.#

This latter is ofc a separate issue for a separate repo, but writing out here to give context.

@adecler @al-fisher @alelom @FraserGreenroyd

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

 Closes #4 

 

Adding method for Push-Pull-Compare for adapter testing

- Adding diffing results to be stored and compared between master and PR branches
- Adding AdapterId based EqualityComparer to make use of venn diagrams to find matching objects

 ### Test files

https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/ETABS_Toolkit/_Macro/Push/ETABS%20-%20Push%20-Pull%20-%20Compare.gh?csf=1&e=U4z4V5

 ### Changelog

 ### Additional comments

### CODE REVIEW

File: InteropabilityTest_Engine/InteropabilityTest_Engine.csproj [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      <HintPath>..\..\BHoM_Engine\Build\Library_Engine.dll</HintPath>
+    </Reference>
+    <Reference Include="Reflection_Engine">
+      <HintPath>..\..\BHoM_Engine\Build\Reflection_Engine.dll</HintPath>
+    </Reference>
+    <Reference Include="Reflection_oM">
+      <HintPath>..\..\BHoM\Build\Reflection_oM.dll</HintPath>
+    </Reference>
+    <Reference Include="Structure_oM">
+      <HintPath>..\..\BHoM\Build\Structure_oM.dll</HintPath>
```
Comment (FraserGreenroyd) [inline InteropabilityTest_Engine/InteropabilityTest_Engine.csproj]: Is this a necessary reference for the project? Happy if it is, just double checking

File: InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        /***************************************************/
+        /**** Constructors                              ****/
+        /***************************************************/
+
+        public AdapterIdComparer(string adapterId)
+        {
+            m_adapterId = adapterId;
+        }
+
```
Comment (FraserGreenroyd) [inline InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs]: Remove blank line

File: InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            object obj2Id;
+
+            //Check if ID exists
+            if (!obj1.CustomData.TryGetValue(m_adapterId, out obj1Id))
+                return false;
+
+            if (!obj2.CustomData.TryGetValue(m_adapterId, out obj2Id))
+                return false;
+
+
```
Comment (FraserGreenroyd) [inline InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs]: Remove extra blank line

File: InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            if (Object.ReferenceEquals(obj, null)) return 0;
+
+            //Check if identifier exists
+            object id;
+
+            if (!obj.CustomData.TryGetValue(m_adapterId, out id))
+                return 0;
+
+            return id.ToString().GetHashCode();
+
```
Comment (FraserGreenroyd) [inline InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs]: Remove extra blank line

File: InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+
+        /***************************************************/
+        /**** Private Fields                            ****/
+        /***************************************************/
+
+        private string m_adapterId;
+
```
Comment (FraserGreenroyd) [inline InteropabilityTest_Engine/Objects/EqualityComparers/AdapterIdComparer.cs]: Remove extra blank line

File: InteropabilityTest_Engine/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("InteropabilityTest_Engine")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("InteropabilityTest_Engine")]
+[assembly: AssemblyCopyright("Copyright ©  2020")]
```
Comment (FraserGreenroyd) [inline InteropabilityTest_Engine/Properties/AssemblyInfo.cs]: Copyright should be for BHoM per other toolkits

File: Test_Engine/Test_Engine.csproj [hunk 7]
```diff
@@ -101,6 +109,14 @@
...
     <Reference Include="System.Net.Http" />
     <Reference Include="System.Xml" />
+    <Reference Include="Testing_Engine">
+      <HintPath>..\..\BHoM_Engine\Build\Testing_Engine.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="Testing_oM">
+      <HintPath>..\..\BHoM\Build\Testing_oM.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
```
Comment (FraserGreenroyd) [inline Test_Engine/Test_Engine.csproj]: Testing_Engine and Testing_oM in those folders no longer exists - they were removed from BHoM and BHoM_Engine in favour of this toolkit - is this the correct references?

File: Test_oM/Results/DiffingResults.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace BH.oM.Test.Results
+{
+    public class DiffingResult : BHoMObject
+    {
+        /***************************************************/
+        /**** Public Properties                         ****/
+        /***************************************************/
+
+        public bool IsEqual { get; set; }
```
Comment (FraserGreenroyd) [inline Test_oM/Results/DiffingResults.cs]: Can we have an explicit default on this property

File: Test_oM/Results/PropertyDifference.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BH.oM.Base;
+
```
Comment (FraserGreenroyd) [inline Test_oM/Results/PropertyDifference.cs]: Remove extra blank line

File: InteropabilityTest_Engine/Compute/PushPullCompare.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                return new List<PushPullSetDiffing>();
+
+            //Get testdata
+            var testSets = Query.TestDataOfType(type);
+            List<string> testSetNames = testSets.Item1;
+            List<List<IBHoMObject>> testData = testSets.Item2;
+
+            //Set up comparer and request
+            FilterRequest request = new FilterRequest { Type = type };
+            AdapterIdComparer comparer = new AdapterIdComparer(adapter.AdapterId);
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply InteropabilityTest_Engine/Compute/PushPullCompare.cs]: This was excluded from the level 04 refactoring. But one to keep an eye on for further refactorings. I am aware of it! 👍

[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply InteropabilityTest_Engine/Compute/PushPullCompare.cs]: image

Are you sure, that's from the instructions for the Lvl04 refactoring?

File: InteropabilityTest_Engine/InteropabilityTest_Engine.csproj (hunk 1 above)
Comment (IsakNaslundBh) [reply InteropabilityTest_Engine/InteropabilityTest_Engine.csproj]: Should not be needed for this PR, but @JonathanNillius will need it for later PRs so might as well leave it in I think

Comment (FraserGreenroyd) [reply InteropabilityTest_Engine/InteropabilityTest_Engine.csproj]: Ok that's fair, happy with that then, resolved

File: InteropabilityTest_Engine/Properties/AssemblyInfo.cs (hunk 6 above)
Comment (IsakNaslundBh) [reply InteropabilityTest_Engine/Properties/AssemblyInfo.cs]: 👍 Took the liberty to update it in the other projects as well, as well as bumping their assembly file version

File: InteropabilityTest_Engine/Compute/PushPullCompare.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply InteropabilityTest_Engine/Compute/PushPullCompare.cs]: Oh, you are right! Sorry. Even so, would prefer to merge this in so jonathan can get started with some other things, Happy to raise a refactoring branch straight after this has been merged.

[the remark this answers is not in the document]
Comment (FraserGreenroyd) [reply InteropabilityTest_Engine/Compute/PushPullCompare.cs]: Yeah I'm not suggesting we hold up this PR for it as I know there's still a couple of weeks to go before refactoring is complete. I'll get an issue raised now so we don't forget to do it :smile:

**Pull request merged**